### ISSUE TITLE
CompletionHandlerException: Exception in resume onCancellation handler for `CancellableContinuation(DispatchedCont`…

### ISSUE THREAD

**Issue by AliSoftware** [wordpress-mobile/WordPress-Android#17463]
**Sentry Url**: https://sentry.io/share/issue/86e0889187bf49408f1f7460cdb241a8/
**User Count**: 281
**Count**: 617
**First Release**: 21.0-rc-1
**First Seen**: 2022-10-18T06:34:51.360000Z
**Last Seen**: 2022-11-15T08:51:07Z
**24 Hours**: 71
**30 Days**: 614

```
CompletionHandlerException: Exception in resume onCancellation handler for CancellableContinuation(DispatchedContinuation[Dispatchers.IO, Continuation at org.wordpress.android.ui.uploads.UploadStarter$upload$2.invokeSuspend(UploadStarter.kt:148)@cc5a5b9]){Completed}@ebc1dfe

...
(1 additional frame(s) were not displayed)
```

**Comment by ovitrif** [comment]
I couldn't reproduce this issue crash, what I've tried is:

1. Open App
2. Disable internet connectivity
2. Create and edit a couple of posts and pages, some to be published some saved as drafts
3. Either
   a. Put app in background
   b. Close app
5. Enable internet connectivity
6. Either
   a. Put app in foreground
   b. Open app

I've tried these steps about 5 times, both using emulator and physical (Samsung) device.

The failure occurs in `UploadStarter.kt:148`, and it's caused by `IllegalStateException` on `mutex.unlock()` based on the stack trace in Sentry.

According to the timing of events reported in Sentry, this happens shortly after the app is started.
Without being able to reproduce it, I can't be sure if this results in data loss (though it might have the potential to), or if the issue recovers automatically with additional attempts to upload their local data.

I traced back the changes to this commit: https://github.com/wordpress-mobile/WordPress-Android/commit/43949f11e88fe29b3710a7812f6c1e03fc18f2fc.

@ParaskP7 since you may have more context on this I tagged you on this one 🙇 .

My understanding is that formerly there was an issue on Samsung device with their fork of Android 5.0, but now we don't support that Android version anymore. I've checked to see if this issue is somehow still linked to Samsung devices, while it occured in a few instances on those, there are other devices affected.

I've triaged this to HIGH priority with a bit of skepticism, since it might auto recover and not necessarily result in data loss (this is based on the fact that I didn't see it repeat for the same user in Sentry).

**Comment by ParaskP7** [comment]
👋 @ovitrif and thanks for this investigation! 🥇 

> I couldn't reproduce this issue crash

Yea, during testing, I couldn't make it crash as well, that is, when we replaced `@Synchronized suspend` with `suspend mutex`.

> My understanding is that formerly there was an issue on Samsung device with their fork of Android 5.0, but now we don't support that Android version anymore. I've checked to see if this issue is somehow still linked to Samsung devices, while it occurred in a few instances on those, there are other devices affected.

Yea, but I think the previous issue was different to this one, see issue [here](https://github.com/wordpress-mobile/WordPress-Android/issues/10827) and fix PR [here](https://github.com/wordpress-mobile/WordPress-Android/pull/10877/). As such I believe this is a different issue. Cc @malinajirka wdyt? 🤔

> According to the timing of events reported in Sentry, this happens shortly after the app is started.
Without being able to reproduce it, I can't be sure if this results in data loss (though it might have the potential to), or if the issue recovers automatically with additional attempts to upload their local dat
> 
> I've triaged this to HIGH priority with a bit of skepticism, since it might auto recover and not necessarily result in data loss.

Thanks for adding the `HIGH` priority, I think this is the correct priority as we should make sure that this doesn't results in data loss.

> (this is based on the fact that I didn't see it repeat for the same user in Sentry)

I am actually seeing it repeat for the same user, see Discover view [here](https://sentry.io/organizations/a8c/discover/results/?field=title&field=release&field=environment&field=user.display&field=timestamp&name=CompletionHandlerException%3A+Exception+in+resume+onCancellation+handler+for+CancellableContinuation%28DispatchedContinuation%5BDis...&project=1438088&query=issue%3AWORDPRESS-ANDROID-2C45&sort=-timestamp&statsPeriod=90d&yAxis=count%28%29).

> I traced back the changes to this commit: https://github.com/wordpress-mobile/WordPress-Android/commit/43949f11e88fe29b3710a7812f6c1e03fc18f2fc.
> 
> @ParaskP7 since you may have more context on this I tagged you on this one 🙇 .

I am not a `Mutex` expert, thus, I am not sure what is happening here, this needs a bit of a debugging and further investigation to get to the bottom of it, along with a way to replicate it first. I am seeing `Pixel 6` crashes as well, which give me some more hope that we could replicate this somehow. 🤞 

Now, I went and read the documentation on [Mutex.lock()](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.sync/-mutex/lock.html) and I am still trying to get my head around it, for example, I still don't (perfectly) understand this below:

```
- This suspending function is cancellable.
- If the `Job` of the current coroutine is cancelled or completed while this function is suspended,
this function immediately resumes with `CancellationException`.
- There is a prompt cancellation guarantee.
- If the job was cancelled while this function was suspended, it will not resume successfully.
- This function releases the lock if it was already acquired by this function before the
`CancellationException` was thrown.
```

`Parameters (owner)`
```
Optional owner token for debugging. When owner is specified (non-null value) and
this mutex is already locked with the same token (same identity),
this function throws `IllegalStateException`.
```

I am a bit confused by all the above... @ovitrif and @malinajirka, what is your take on what documentation is saying? Any ideas on how to progress from here? 🤔

**Comment by malinajirka** [comment]
Looking at the crash, it mentions `Mutex is not locked` => I haven't done a proper research, but my personal guess is that the coroutine gets cancelled -> the system automatically unlocks the mutex as per `If the `Job` of the current coroutine is cancelled or completed while this function is suspended,
this function immediately resumes with `CancellationException`.` => we attempt to unlock it again in `finally` => 💥 . (even though my gut feeling tells me that this shouldn't happen as I'd expect the `finally` to be processed before the cancellation 🤷 )

I'd personally try to reproduce this by modifying the code - or even writing a dummy code with the same setup (Lock, cancel, unlock in finally) -> observe how it behaves or we could just try wrapping the `unlock` in `if(mutex.isLocked) {...}`. Wdyt?

**Comment by ParaskP7** [comment]
> Looking at the crash, it mentions Mutex is not locked => I haven't done a proper research, but my personal guess is that the coroutine gets cancelled -> the system automatically unlocks the mutex as per If the Jobof the current coroutine is cancelled or completed while this function is suspended, this function immediately resumes withCancellationException. => we attempt to unlock it again in finally => 💥 . (even though my gut feeling tells me that this shouldn't happen as I'd expect the finally to be processed before the cancellation 🤷 )

Thank @malinajirka , this is exactly what I am thinking too... 💯 

> I'd personally try to reproduce this by modifying the code - or even writing a dummy code with the same setup (Lock, cancel, unlock in finally) -> observe how it behaves or we could just try wrapping the unlock in if(mutex.isLocked) {...}. Wdyt?

Yes, I like both suggestions, if it were me, I would first go with adding the `if(mutex.isLocked) {...}` to guard the `mutex.unlock()` and include it as quick as possible to the codebase. In the meanwhile, I would have tried to reproduce this somehow, with a dummy code or otherwise. Wdyt @ovitrif ? 🤔

**Comment by ovitrif** [comment]
Thank you @ParaskP7 & @malinajirka for your feedback and insights on this issue 🙇‍♂️ 

I also had on my mind that we could use `if(mutex.isLocked) { mutex.unlock() }` but wasn't sure how much of a proper fix this would be.

As I've understood from your feedback actually the `if(mutex.isLocked)` is a valid fix if this is indeed correct (bonus points if we can prove it):
> the coroutine gets cancelled -> the system automatically unlocks the mutex as per `If the `Job` of the current coroutine is cancelled or completed while this function is suspended,
this function immediately resumes with `CancellationException`.` => we attempt to unlock it again in `finally` => 💥

---

> I am actually seeing it repeat for the same user, see Discover view [here](https://sentry.io/organizations/a8c/discover/results/?field=title&field=release&field=environment&field=user.display&field=timestamp&name=CompletionHandlerException%3A+Exception+in+resume+onCancellation+handler+for+CancellableContinuation%28DispatchedContinuation%5BDis...&project=1438088&query=issue%3AWORDPRESS-ANDROID-2C45&sort=-timestamp&statsPeriod=90d&yAxis=count%28%29).

Thank you for this info @ParaskP7, I could actually see it occur for the same user, and repeatedly while attempting to sync the same post to the server, which increases the likelihood of it causing data loss.

> Yes, I like both suggestions, if it were me, I would first go with adding the if(mutex.isLocked) {...} to guard the mutex.unlock() and include it as quick as possible to the codebase. In the meanwhile, I would have tried to reproduce this somehow, with a dummy code or otherwise.

This makes perfect sense to me, thank you once again 🙇 , I'm not a `mutex` expert myself but I'll give this a try with dummy code and see if I can reproduce it; as for a start I'll already open a PR that guards the `mutex.unlock()` call.

**Comment by ParaskP7** [comment]
👋 @ovitrif !

> As I've understood from your feedback actually the if(mutex.isLocked) is a valid fix if this is indeed correct (bonus points if we can prove it):

💯 

> This makes perfect sense to me, thank you once again 🙇 , I'm not a mutex expert myself but I'll give this a try with dummy code and see if I can reproduce it; as for a start I'll already open a PR that guards the mutex.unlock() call.

Awesome, feel free to add me as a reviewer to this PR that you've already opened, I can't seem to find it myself, I am probably lucking some GitHub search skills here... 😅

**Comment by ovitrif** [comment]
> Awesome, feel free to add me as a reviewer to this PR that you've already opened, I can't seem to find it myself, I am probably lucking some GitHub search skills here... 😅

👋 @ParaskP7 I've just opened the PR now 🙇 :
- https://github.com/wordpress-mobile/WordPress-Android/pull/17472

**Comment by ovitrif** [comment]
Let's keep an eye on this by checking it in Sentry for release `21.2.0.1` before we proceed to close it.

P.S. I couldn't reproduce it with dummy code.

**Comment by thehenrybyrd** [comment]
@ovitrif can you confirm whether this is resolved?

**Comment by ovitrif** [comment]
> @ovitrif can you confirm whether this is resolved?

Thanks for the ping @thehenrybyrd.

I inspected the crash occurrences related to this issue in the [last 30 days in Sentry](https://sentry.io/organizations/a8c/discover/results/?field=title&field=release&field=environment&field=user.display&field=timestamp&name=CompletionHandlerException%3A+Exception+in+resume+onCancellation+handler+for+CancellableContinuation%28DispatchedContinuation%5BDis...&project=1438088&query=issue%3AWORDPRESS-ANDROID-2C45&sort=-timestamp&statsPeriod=30d&yAxis=count%28%29) and it's still reproducing on the latest versions:

<img width="1370" alt="completionHandlerExceptionStats" src="https://user-images.githubusercontent.com/4588074/212037704-28dec825-f01a-49c7-bc60-f48d6023cc2d.png">

I suspect we may have not fully understood the cause of the issue.

From @malinajirka's remark:

> Looking at the crash, it mentions Mutex is not locked => I haven't done a proper research, but my personal guess is that the coroutine gets cancelled -> the system automatically unlocks the mutex as per If the Jobof the current coroutine is cancelled or completed while this function is suspended, this function immediately resumes withCancellationException. => we attempt to unlock it again in finally => 💥 . (even though my gut feeling tells me that this shouldn't happen as I'd expect the finally to be processed before the cancellation 🤷 )

It looks to me that the crash actually occurs at the step in bold:

> the coroutine gets cancelled -> **the system automatically unlocks the mutex**

The difference between the above and what we've tried to fix has to do with the fact that the crash happens in the system code, not in our code. It might be that our code doesn't guard specifically for cases where the system produces the crash, or even that our code might create the conditions for the system to produce the crash.

My suggestion is to keep investigating this issue during groundskeeping, and keep it as a high priority since it could result in data loss.

**Comment by ParaskP7** [comment]
> My suggestion is to keep investigating this issue during groundskeeping, and keep it as a high priority since it could result in data loss.

I agree @ovitrif @thehenrybyrd ! 💯

**Comment by zwarm** [comment]
@ParaskP7 @ovitrif 
What if this is a result of a race condition on the `mutex.lock()/unlock()`? Doesn't the `mutex.isLocked()` only check the state in the current moment? How about if we use `mutex.withLock{ ...}`  instead of `mutex.lock()` in the try and then eliminate the `if + mutex.unlock()` check completely in finally? The `mutex.withLock` should automatically lock/unlock the mutex for us. I think it's worth a shot. wdyt?

**Comment by ovitrif** [comment]
> What if this is a result of a race condition on the `mutex.lock()/unlock()`?

It's quite hard for me to understand the internals of the mutex system when I'm not fully understanding the internals of coroutines(cancellable and continuations) with relations to a mutex.
When I was investigating this I couldn't manage to reproduce it by running the app with the debugger attached neither by writing unit tests for it.

TBH I'm not sure how a race condition might influence the exception, since as I [mentioned in my previous comment](https://github.com/wordpress-mobile/WordPress-Android/issues/17463#issuecomment-1380094365) my current understanding based on the stack trace we get in Sentry is that <ins>the mutex is not locked when the coroutine internals calls `unlock()` on it</ins>. I never found any any reference to the app's `mutex.unlock()` call in the stack traces from Sentry.

> Doesn't the `mutex.isLocked()` only check the state in the current moment?

This issue started before I added the `isLocked()` check. Back then `mutex.unlock()` was the only code in the `finally` block. Because of this I believe the fault is not related to how isLocked runs, plus `unlock()` is not async, so in theory it runs in the same current moment as well.

There is also no reference in any stack trace to the `unlock()` line of code.

> How about if we use `mutex.withLock{ ...}` instead of `mutex.lock()` in the try and then eliminate the `if + mutex.unlock()` check completely in finally? The `mutex.withLock` should automatically lock/unlock the mutex for us. I think it's worth a shot. wdyt?

Good point, I had the same idea but my enthusiasm fade after I read its code:
```kt
// contract ...
lock(owner)
try {
    return action()
} finally {
    unlock(owner)
}
```
It's almost the same to what we did initially, except the `lock()` is before the `try`.
I wouldn't expect it makes a significant difference.

---

@zwarm If your idea is clear enough to write a failing unit tests for it I'd be happy to start from there, or if you could elaborate a little more I could try to write the test myself.

**Comment by ovitrif** [comment]
> > What if this is a result of a race condition on the mutex.lock()/unlock()?
>
> I'm not sure how a race condition might influence the exception

Actually I do think this might be plausible:

---

### ➀ Race condition scenario

1. First `upload()` starts, calling the async `mutex.lock()`
   - A job is started and keeps running
2. Second `upload()` starts, calling `lock()` on the same mutex instance
3. Second `lock()` acquires the lock
4. First `upload()` completes, calling the sync `mutex.unlock()`
5. The Second `upload()` gets cancelled 
   - 💣 At this point it tries to release the lock throwing `Mutex is not locked`

---

@zwarm @ParaskP7 (and if I may @malinajirka 🙇🏻 ) wdyt, does the above scenario make sense to you as well?

**Comment by zwarm** [comment]
@ovitrif 
> Actually I do think this might be plausible: 
> ➀ Race condition scenario

That's exactly what I was thinking. Also thinking about those `async` calls. I don't think they hold onto the mutex lock, but it's been a while and I could be remembering incorrectly. I'm not all that familiar with what is going on here and/or which pieces need to hold onto the lock to execute. And I agree with @ovitrif , starting out with a failing test is a good idea. I'll give it some more thought as the team progresses through our release/support rotation.

**Comment by ParaskP7** [comment]
👋 @zwarm @ovitrif !

I am following the discussions you are having (plus #18067) and I think you are on the correct track here, especially on first trying to identify the race condition scenario. Unfortunately, I don't have much to suggest to you apart from going through trial and error in figuring out how to replicate this crash first.

> Good point, I had the same idea but my enthusiasm fade after I read its code:

Yeaaa... 🤷

**Comment by ovitrif** [comment]
Merged this PR:
- https://github.com/wordpress-mobile/WordPress-Android/pull/18067

I propose to go ahead and check if it fixes the issue, GPT convinced me it might actually be a solution. Let's see if I get disappointed again by AI :trollface:, it wouldn't be the first time…

**Comment by ovitrif** [comment]
The first thing I'm noticing that feels a bit unnatural as a developer is the lack of unit tests for a public method that could not only run the code where the crash occurs but also in what we believe is the right environment to cause it, ie. "multiple jobs of multiple sites uploading in parallel".

Good luck finding the unit test running this code:

![CleanShot 2023-05-16 at 14 45 43](https://github.com/wordpress-mobile/WordPress-Android/assets/4588074/d9ba7e93-7e1d-42e5-bb22-683bc7ea3fe9)

**Comment by ovitrif** [comment]
Removing my assignment since I'm no longer focusing on Maintenance. I am still available to assist the brave one with knowledge transfer from my learnings and discoveries while attempting to repro this crash via TDD in:
- #18449

**Comment by mkevins** [comment]
Marking this as resolved, since:

The coroutines version has been bumped (https://github.com/wordpress-mobile/WordPress-Android/commit/1f2c2d3884d3040c5e150c71075db7a4db4a1f04), which landed in release 22.7, which [notably](https://github.com/wordpress-mobile/WordPress-Android/pull/18449#pullrequestreview-1431533409) changes the Mutex implementation, and [likely resolves this issue](https://github.com/wordpress-mobile/WordPress-Android/pull/18449#discussion_r1197885047).

**Comment by marcprux** [comment]
I just hit this crash on a Pixel 6 (Android 13) with `kotlinx-coroutines-android:1.7.3`.

```
11-20 00:24:58.233 10433 10433 E AndroidRuntime: kotlinx.coroutines.CompletionHandlerException: Exception in resume onCancellation handler for CancellableContinuation(DispatchedContinuation[AndroidUiDispatcher@444598f, Continuation at androidx.compose.foundation.gestures.PressGestureScopeImpl.reset(TapGestureDetector.kt:357)@7ef7dfa]){Completed}@2b6ffab
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at kotlinx.coroutines.CancellableContinuationImpl.callOnCancellation(CancellableContinuationImpl.kt:264)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at kotlinx.coroutines.CompletedContinuation.invokeHandlers(CancellableContinuationImpl.kt:659)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at kotlinx.coroutines.CancellableContinuationImpl.cancelCompletedResult$kotlinx_coroutines_core(CancellableContinuationImpl.kt:180)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:102)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at androidx.compose.ui.platform.AndroidUiDispatcher.performTrampolineDispatch(AndroidUiDispatcher.android.kt:81)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at androidx.compose.ui.platform.AndroidUiDispatcher.access$performTrampolineDispatch(AndroidUiDispatcher.android.kt:41)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.run(AndroidUiDispatcher.android.kt:57)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:942)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:201)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:288)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:7918)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [androidx.compose.ui.platform.MotionDurationScaleImpl@eb89ab4, androidx.compose.runtime.BroadcastFrameClock@ef4e6dd, StandaloneCoroutine{Cancelling}@88bd252, AndroidUiDispatcher@444598f]
11-20 00:24:58.233 10433 10433 E AndroidRuntime: Caused by: java.lang.IllegalStateException: This mutex is not locked
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at kotlinx.coroutines.sync.MutexImpl.unlock(Mutex.kt:213)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at kotlinx.coroutines.sync.MutexImpl$CancellableContinuationWithOwner$tryResume$token$1.invoke(Mutex.kt:261)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at kotlinx.coroutines.sync.MutexImpl$CancellableContinuationWithOwner$tryResume$token$1.invoke(Mutex.kt:258)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	at kotlinx.coroutines.CancellableContinuationImpl.callOnCancellation(CancellableContinuationImpl.kt:259)
11-20 00:24:58.233 10433 10433 E AndroidRuntime: 	... 14 more
```
